### ISSUE TITLE
prefer_interpolation_to_compose_strings should not apply if operands are not on the same line

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#2668]
If the operands are not on the same line the change could lead to really long line exceeding the max line length.

According to the current lint:

```dart
  var a = a_________________really______________long_______________string__________expression +
      '  ' +
      another_________________really______________long_______________string__________expression;
```

should be changed to:

```dart
  var a = 
      '${a_________________really______________long_______________string__________expression}  ${another_________________really______________long_______________string__________expression}';
```

But the result is really bad.

**Comment by bwilkerson** [comment]
My recommendation would _not_ be to leave the code as is, but to transform the code into
```
var first = a_________________really______________long_______________string__________expression;
var second = another_________________really______________long_______________string__________expression;
var a = '$first $second';
```

Do we really want to not recommend using interpolation, or do we also want to recommend shortening long interpolations by extracting expressions into local variables?

**Comment by a14n** [comment]
I agree that extracting a variable from string interpolation can sometimes improve readability when the line length exceeds a max value. However it's not directly related to this lint (IMHO it's more related to the `line_length` rule). Moreover naming variables is one of the hardest problem in writing code :)

**Comment by bwilkerson** [comment]
I'm still not convinced that we want to promote the style of the first example, nor am I convinced that we want lints to be sensitive to line length (or worse yet, each other). There's much to be said for lints being independent and easy to understand (without a lot of special cases). If, because of line length, it's ok to write
```
  var a = a_________________really______________long_______________string__________expression +
      '  ' +
      another_________________really______________long_______________string__________expression;
```
but not ok to write
```
  var a = an_expression + ' ' + another_expression;
```
then that means that there's a single character threshold somewhere where the applicability of the lint switches on and off, and I would find that to be annoying.

**Comment by a14n** [comment]
Here's an example from flutter where the current code looks better without string composition (and even without introducing a new variable)

```dart
    final String build = testEnvironment.buildDir.path;
    processManager.addCommands(<FakeCommand>[
      FakeCommand(command: <String>[
        artifacts.getHostArtifact(HostArtifact.engineDartBinary).path,
        '--disable-dart-dev',
        artifacts.getArtifactPath(Artifact.frontendServerSnapshotForEngineDartSdk),
        '--sdk-root',
        artifacts.getArtifactPath(
          Artifact.flutterPatchedSdkPath,
          platform: TargetPlatform.android_arm,
          mode: BuildMode.debug,
        ) + '/',                                                    // LINT HERE
        '--target=flutter',
        '--no-print-incremental-dependencies',
        ...buildModeOptions(BuildMode.debug, <String>[]),
        '--track-widget-creation',
        '--no-link-platform',
        '--packages',
        '/.dart_tool/package_config.json',
        '--output-dill',
        '$build/app.dill',
        '--depfile',
        '$build/kernel_snapshot.d',
        'file:///lib/main.dart',
      ], stdout: 'result $kBoundaryKey\n$kBoundaryKey\n$kBoundaryKey /build/653e11a8e6908714056a57cd6b4f602a/app.dill 0\n'),
    ]);
```

**Comment by a14n** [comment]
@bwilkerson if we don't special case multiline concatenation with `+` how would you refactor the following code to avoid the lint:

```dart
final String path = artifacts.getArtifactPath(
      Artifact.flutterPatchedSdkPath,
      platform: TargetPlatform.android_arm,
      mode: BuildMode.debug,
    ) +
    '/';
```

**Comment by bwilkerson** [comment]
I would have written it as
```
final String path = artifacts.getArtifactPath(
    Artifact.flutterPatchedSdkPath,
    platform: TargetPlatform.android_arm,
    mode: BuildMode.debug,
  );
...
      '--sdk-root',
      '$path/',
      '--target=flutter',
...
```
I think the list of command arguments looks much better when each argument occupies a single line.

**Comment by a14n** [comment]
It's ok if `path` is used only once but this leads to duplications if it is used at several places (actually there are several similar cases). One additional issue is that `path` logic is now splitted and you have to deal with trailing slash at 2 sites (instead of defining a variable that ends with slash)

**Comment by bwilkerson** [comment]
If `path` is used in multiple places, or if I was concerned about splitting the path logic across expressions (which is valid), I'd define a private method to get the sdk path and write
```
final String sdkPath = _getSdkPath();
```
and the only reason I wouldn't inline the call is because of the duplication.

I don't think it's readable to use a binary operator when either of the operands is too long (and too long to fit on one line certainly qualifies, though even shorter expressions might as well), and introducing local variables or utility methods in order to make the code more readable is a positive tradeoff in my esthetic.

**Comment by a14n** [comment]
Thanks for this discussion. After some refactoring it is not as bad as I thought it would be.
